Smartarray API improvements.

stefanseefeld opened the pull request:
This PR generalizes the SmartArray API by replacing methods `host()` and `gpu()` by `get(where)`, and likewise replaces `host_changed()` and `gpu_changed()` by `mark_changed(where)`. (Thus there is nothing GPU-specific in the API any longer.)
The PR further adds support for NumPy's `astype` functionality, which we use elsewhere for type promotion. (How to deal with type promotion in device memory has yet to be decided...)

stuartarchibald commented on numba/tests/test_smart_array.py:
```diff
@@ -57,8 +59,14 @@ def test_ufunc(self):
 
+    def test_astype(self):
+        a = SmartArray(np.int32([42, 8, -5]))
+        #cfunc = jit(nopython=True)(astype_usecase)
```
Is this supposed to be here?

stefanseefeld replied:
oups, forgot to reenable (after fixing the pythonapi bug). Thanks for catching that.

stuartarchibald commented on numba/smartarray.py:
```diff
@@ -92,33 +92,27 @@ def ndim(self): return self._ndim
+            self._invalidate('host')
+            # only sync if there are active views
+            if self._host is not None and sys.getrefcount(self._host) > 2:
+                self._sync('host')
```
and if `where=='anythingelse'` what will happen?

stefanseefeld replied:
I shall raise a ValueError throughout that module if 'where' is not a valid target.

stuartarchibald commented on numba/tests/test_smart_array.py:
```diff
@@ -23,6 +23,8 @@ def shape_usecase(x):
 def npyufunc_usecase(x):
     return np.cos(np.sin(x))
 
+def astype_usecase(x, dtype):
```
Is this dead code?

stefanseefeld replied:
Yes, and thus removed.

stuartarchibald commented on numba/tests/test_smart_array.py:
```diff
@@ -57,8 +59,13 @@ def test_ufunc(self):
+        a = SmartArray(np.int32([42, 8, -5]))
+        aa = a.astype(np.float64)
+        self.assertIsInstance(aa, SmartArray)
+        self.assertPreciseEqual(aa.get('host'), a.get('host').astype(np.float64))
```
Perhaps check the `dtype` is correct through comparison to a hardcoded dtype, else this is just checking that the function is working/broken in the same way?

stefanseefeld replied:
Arguably that's what we should check: that the call is properly forwarded to the nested NDArray.

pitrou replied:
Indeed, you should mutate the `astype()` return and check that it mutates the original array.

seibert commented:
I think we should continue to support the old API with a deprecation warning for a few releases, in case someone is using it.  We marked it as experimental, but I don't want to break someone's code with no warning.

Also, there should be corresponding documentation changes to match up with the API change.

stefanseefeld commented:
I have re-enabled the old API (and added a deprecation API to mark it properly). I have also updated the documentation. (I suppose it's not necessary to keep the documentation of the old API.)

stuartarchibald commented on numba/errors.py:
```diff
@@ -7,11 +7,24 @@
 # Filled at the end
 __all__ = []
 
+def deprecated(subst):
```
Docstring?

stuartarchibald replied:
Also, what if there's no substitution and something is just being removed?

stuartarchibald commented on numba/tests/test_warnings.py:
```diff
@@ -107,5 +106,18 @@ def do_loop(x):
+        @deprecated('foo')
+        def bar(): pass
+
+        with warnings.catch_warnings(record=True) as w:
```
Think this block can be done in 1 line with an `assertRaisesRegex` and an `error` filter?

stefanseefeld commented:
The deprecated function now makes the parameter optional, and contains some docs.

pitrou commented on numba/errors.py:
```diff
@@ -7,11 +7,40 @@
+
+        return wraps(func)(wrapper)
+
+    if callable(arg) or type(arg) in (classmethod, staticmethod):
```
`if subst is None` sounds better and simpler here.

pitrou commented on numba/smartarray.py:
```diff
@@ -92,33 +96,41 @@ def ndim(self): return self._ndim
+        if where == 'host':
+            self._invalidate('gpu')
+            # only sync if there are active views
+            if self._gpu is not None and sys.getrefcount(self._gpu) > 2:
```
The reliance on `getrefcount` looks fragile here. What if another reference is always kept somewhere else?

stefanseefeld replied:
If anyone refers to the array it needs to be synced, I don't see any way around that.

pitrou replied:
I mean getrefcount() does not sound a very reliable way of knowing whether the array is referenced by other views. Also, I'm not sure whether you need to sync here or only invalidate the views.

stefanseefeld replied:
Is there a way to reference the array without increasing the ref count ?

pitrou replied:
With a weakref in Python, or a raw pointer in C. However, my concern is rather for the reverse situation: if the reference count is increased for other reasons, this method would synchronize the array even though it's not needed.

stefanseefeld replied:
True. Unfortunately I'm not aware of any other way to measure this. (Ideally we'll eventually "fix" the NumPy C API so one can't access the raw pointer any longer, precisely so it becomes possible to encapsulate the data storage. One step at a time...

pitrou commented on numba/targets/boxing.py:
```diff
@@ -338,12 +338,14 @@ def box_smart_array(typ, value, c):
         with has_parent:
             c.pyapi.incref(obj)
-            retn = c.pyapi.call_method(obj, "host_changed")
+            host = c.pyapi.string_from_constant_string('host')
```
Are there tests covering this? It seems the mistake in `pythonapi.call_method()` went unnoticed previously...

stefanseefeld replied:
Are there any tests anywhere for the pyapi module ? The error went indeed unnoticed, because we only called this method without arguments.

pitrou replied:
Oh, I see. No, the pyapi module doesn't have unit tests. They would not be very practical to write.

stefanseefeld replied:
So I'm not quite sure what you are proposing.

seibert commented:
@pitrou: If you are happy with the code review, feel free to merge this PR.

pitrou commented:
There are a couple comments remaining that would need addressing first (though nothing severe).

stefanseefeld commented:
I believe I have addressed all comments, so this PR should be ready to be merged.

pitrou commented:
I think you forgot to address the comments about improving the test case.
